compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -201,19 +202,26 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
 
         // We may have to repeatedly recompute the goal in case of coinductive cycles,
         // check out the `cache` module for more information.
-        //
-        // FIXME: Similar to `evaluate_all`, this has to check for overflow.
-        loop {
-            let (ref infcx, goal, var_values) =
-                tcx.infer_ctxt().build_with_canonical(DUMMY_SP, &canonical_goal);
-            let mut ecx =
-                EvalCtxt { infcx, var_values, search_graph, in_projection_eq_hack: false };
-            let result = ecx.compute_goal(goal);
-
-            if search_graph.try_finalize_goal(tcx, canonical_goal, result) {
-                return result;
-            }
-        }
+        search_graph.repeat_while_none(
+            |this| this.deal_with_overflow(tcx, canonical_goal).unwrap(),
```

Comment:
instead of using `unwrap()` here only to wrap it again inside of `repeat_while_none`, change `repeat_while_none` to return `T` as well I think.

Thinking about this I think we shouldn't merge this without correctly cleaning up the stack in the overflow case. Right now we don't pop the current goal from the stack.

To do that I think we should change the layout of the code here:

change `fn evaluate_canonical_goal` to the following

```rust
    fn evaluate_canonical_goal(
        tcx: TyCtxt<'tcx>,
        search_graph: &'a mut search_graph::SearchGraph<'tcx>,
        canonical_goal: CanonicalGoal<'tcx>,
    ) -> QueryResult<'tcx> {
        // Deal with overflow, caching, and coinduction.
        //
        // The actual solver logic happens in `ecx.compute_goal`.
        search_graph.with_new_goal(
            tcx,
            canonical_goal,
            |search_graph| {
                let (ref infcx, goal, var_values) =
                tcx.infer_ctxt().build_with_canonical(DUMMY_SP, &canonical_goal);
                let mut ecx = 
                    EvalCtxt { infcx, var_values, search_graph, in_projection_eq_hack: false };
                ecx.compute_goal(goal)
            }
        )
    }
```

merging `try_push_stack` the loop for coinduction and `try_finalize_goal` into a new function `search_graph.with_new_goal`. There might be a better name than `fn with_new_goal` :shrug: